File: macos/Sources/DesktopVoiceInput/Services/Providers/ProviderFactory.swift
import Foundation

@MainActor
final class ProviderFactory {
    private unowned let settings: AppSettings

    init(settings: AppSettings) {
        self.settings = settings
    }

    func resolveProviders() -> [ProviderSelection] {
        let config = settings.recognitionConfig

        switch config.mode {
        case .auto:
            var providers: [ProviderSelection] = [ProviderSelection(mode: .local, provider: LocalSpeechProvider())]

            if config.doubaoCredentials.isConfigured {
                providers.append(ProviderSelection(mode: .doubao, provider: DoubaoSpeechProvider()))
            }

            if config.qwenCredentials.isConfigured {
                providers.append(ProviderSelection(mode: .qwen, provider: QwenSpeechProvider()))
            }

            return providers
        case .local:
            return [ProviderSelection(mode: .local, provider: LocalSpeechProvider())]
        case .doubao:
            return [ProviderSelection(mode: .doubao, provider: DoubaoSpeechProvider())]
        case .qwen:
            return [ProviderSelection(mode: .qwen, provider: QwenSpeechProvider())]
        }
    }
}

File: windows/src/GuGuTalk.Core/Providers/ProviderFactory.cs
using GuGuTalk.Core.Models;
using GuGuTalk.Core.Settings;

namespace GuGuTalk.Core.Providers;

public sealed class ProviderFactory
{
    private readonly AppSettings _settings;

    public ProviderFactory(AppSettings settings)
    {
        _settings = settings;
    }

    public List<ProviderSelection> ResolveProviders()
    {
        var selections = new List<ProviderSelection>();

        switch (_settings.PreferredMode)
        {
            case RecognitionMode.Local:
                // Local provider added by app layer (GuGuTalk.LocalAsr)
                break;
            case RecognitionMode.Doubao:
                if (_settings.DoubaoCredentials.IsConfigured)
                    selections.Add(new ProviderSelection(RecognitionMode.Doubao, new DoubaoSpeechProvider()));
                if (_settings.QwenCredentials.IsConfigured)
                    selections.Add(new ProviderSelection(RecognitionMode.Qwen, new QwenSpeechProvider()));
                break;
            case RecognitionMode.Qwen:
                if (_settings.QwenCredentials.IsConfigured)
                    selections.Add(new ProviderSelection(RecognitionMode.Qwen, new QwenSpeechProvider()));
                if (_settings.DoubaoCredentials.IsConfigured)
                    selections.Add(new ProviderSelection(RecognitionMode.Doubao, new DoubaoSpeechProvider()));
                break;
        }

        return selections;
    }
}
